File: src/main/java/de/bissell/easyclock/ClockFrame.java
package de.bissell.easyclock;

import javax.swing.*;
import javax.swing.border.EmptyBorder;
import java.awt.*;
import java.awt.event.*;
import java.text.SimpleDateFormat;
import java.util.Date;

public class ClockFrame extends JFrame implements ActionListener {

    public static final String CLOCK_FONT = "Courier New";

    private final JLabel label;
    private Settings settings;

    public ClockFrame() {
        settings = Settings.loadSettings();

        label = createLabel();
        label.setText(getTimeString());
        getContentPane().add(label);

        setDefaultCloseOperation(WindowConstants.EXIT_ON_CLOSE);
        setUndecorated(true);
        pack();
        Rectangle bounds = getBounds();
        setBounds(settings.x, settings.y, bounds.width, bounds.height);

        MouseAdapter adapter = new ClockMouseAdapter();
        addMouseListener(adapter);
        addMouseMotionListener(adapter);
        addMouseWheelListener(adapter);

        Timer timer = new Timer(100, this);
        timer.setRepeats(true);
        timer.start();
    }

    private JLabel createLabel() {
        Font font = new Font(CLOCK_FONT, Font.BOLD, settings.fontSize);

        JLabel label = new JLabel();

        label.setHorizontalAlignment(JLabel.CENTER);
        label.setFont(font);
        label.setForeground(Color.red);
        label.setBackground(Color.black);
        label.setOpaque(true);
        label.setBorder(new EmptyBorder(10, 10, 10, 10));

        return label;
    }

    @Override
    public void actionPerformed(ActionEvent e) {
        label.setText(getTimeString());
    }

    private String getTimeString() {
        return new SimpleDateFormat("HH:mm").format(new Date());
    }

    private void resizeLabelFont(int diff) {
        settings.fontSize = Math.max(10, settings.fontSize + diff * 5);
        label.setFont(new Font(CLOCK_FONT, Font.BOLD, settings.fontSize));
        pack();
    }

    private void saveSettingsAndExit() {
        settings.save();

        System.exit(0);
    }

    private class ClockMouseAdapter extends MouseAdapter {

        private Point dragStart = null;

        @Override
        public void mousePressed(MouseEvent e) {
            dragStart = new Point(e.getX(), e.getY());
        }

        @Override
        public void mouseReleased(MouseEvent e) {
            if (e.getButton() == MouseEvent.BUTTON3) {
                saveSettingsAndExit();
            }

            dragStart = null;
        }

        @Override
        public void mouseWheelMoved(MouseWheelEvent e) {
            resizeLabelFont(e.getWheelRotation());
        }

        @Override
        public void mouseDragged(MouseEvent e) {
            Rectangle bounds = getBounds();

            int diffX = e.getX() - dragStart.x;
            int diffY = e.getY() - dragStart.y;

            settings.x = bounds.x + diffX;
            settings.y = bounds.y + diffY;
            setBounds(settings.x, settings.y, bounds.width, bounds.height);
        }
    }
}

File: src/main/kotlin/de/stefanbissell/easyclock/ClockFrame.kt
package de.stefanbissell.easyclock

import de.stefanbissell.easyclock.Settings.Companion.loadSettings
import javax.swing.JFrame
import javax.swing.JLabel
import java.awt.Font
import java.awt.Color
import javax.swing.border.EmptyBorder
import java.text.SimpleDateFormat
import java.awt.Point
import java.awt.event.*
import java.util.*
import javax.swing.Timer
import kotlin.math.max
import kotlin.system.exitProcess

class ClockFrame : JFrame(), ActionListener {

    private val clockFont = "Courier New"

    private val settings = loadSettings()
    private val label = createLabel().apply {
        text = timeString
    }
    private val timeString: String
        get() = SimpleDateFormat("HH:mm").format(Date())

    init {
        contentPane.add(label)
        defaultCloseOperation = EXIT_ON_CLOSE
        isUndecorated = true
        pack()
        setBounds(settings.x, settings.y, bounds.width, bounds.height)
        val adapter = ClockMouseAdapter()
        addMouseListener(adapter)
        addMouseMotionListener(adapter)
        addMouseWheelListener(adapter)
        Timer(100, this).apply {
            isRepeats = true
            start()
        }
    }

    private fun createLabel(): JLabel =
        JLabel().apply {
            horizontalAlignment = JLabel.CENTER
            font = Font(clockFont, Font.BOLD, settings.fontSize)
            foreground = Color.red
            background = Color.black
            isOpaque = true
            border = EmptyBorder(10, 10, 10, 10)
        }

    override fun actionPerformed(e: ActionEvent) {
        label.text = timeString
    }

    private fun resizeLabelFont(diff: Int) {
        settings.fontSize = max(10, settings.fontSize + diff * 5)
        label.font = Font(clockFont, Font.BOLD, settings.fontSize)
        pack()
    }

    private fun saveSettingsAndExit() {
        settings.save()
        exitProcess(0)
    }

    private inner class ClockMouseAdapter : MouseAdapter() {

        private var dragStart: Point? = null
        override fun mousePressed(e: MouseEvent) {
            dragStart = Point(e.x, e.y)
        }

        override fun mouseReleased(e: MouseEvent) {
            if (e.button == MouseEvent.BUTTON3) {
                saveSettingsAndExit()
            }
            dragStart = null
        }

        override fun mouseWheelMoved(e: MouseWheelEvent) {
            resizeLabelFont(e.wheelRotation)
        }

        override fun mouseDragged(e: MouseEvent) {
            val bounds = bounds
            val diffX = e.x - dragStart!!.x
            val diffY = e.y - dragStart!!.y
            settings.x = bounds.x + diffX
            settings.y = bounds.y + diffY
            setBounds(settings.x, settings.y, bounds.width, bounds.height)
        }
    }
}
